Content.Server/Discord/DiscordLink/DiscordLink.cs

```diff
@@ -0,0 +1,257 @@
+
+        _client.Ready += () =>
+        {
+            _sawmill.Info("Discord client ready.");
+            return Task.CompletedTask;
+        };
+
+        Task.Run(async () =>
+        {
+            try
+            {
+                await LoginAsync(token);
+            }
+            catch (Exception e)
+            {
+                _sawmill.Error("Failed to connect to Discord!", e);
+            }
+        });
+    }
+
+    public async Task Shutdown()
+    {
+        if (_client != null)
+        {
+            _sawmill.Info("Disconnecting from Discord.");
+
+            // Unsubscribe from the events.
+            _client.MessageReceived -= OnCommandReceivedInternal;
+            _client.MessageReceived -= OnMessageReceivedInternal;
+
+            await _client.LogoutAsync();
+            await _client.StopAsync();
+            await _client.DisposeAsync();
+            _client = null;
+        }
+
+        _configuration.UnsubValueChanged(CCVars.DiscordGuildId, OnGuildIdChanged);
+        _configuration.UnsubValueChanged(CCVars.DiscordPrefix, OnPrefixChanged);
+    }
+
+    void IPostInjectInit.PostInject()
+    {
+        _sawmill = _logManager.GetSawmill("discord.link");
+        _sawmillNet = _logManager.GetSawmill("discord.link.log");
+    }
+
+    private void OnGuildIdChanged(string guildId)
+    {
+        _guildId = ulong.TryParse(guildId, out var id) ? id : 0;
+    }
+
+    private void OnPrefixChanged(string prefix)
+    {
+        BotPrefix = prefix;
+    }
+
+    private async Task LoginAsync(string token)
+    {
+        DebugTools.Assert(_client != null);
+        DebugTools.Assert(_client.LoginState == LoginState.LoggedOut);
+
+        await _client.LoginAsync(TokenType.Bot, token);
+        await _client.StartAsync();
+
+
+        _sawmill.Info("Connected to Discord.");
+    }
+
+    private string FormatLog(LogMessage msg)
+    {
+        return msg.Exception is null
+            ? $"{msg.Source}: {msg.Message}"
+            : $"{msg.Source}: {msg.Message}\n{msg.Exception}";
+    }
+
+    private Task Log(LogMessage msg)
+    {
+        var logLevel = msg.Severity switch
+        {
+            LogSeverity.Critical => LogLevel.Fatal,
+            LogSeverity.Error => LogLevel.Error,
+            LogSeverity.Warning => LogLevel.Warning,
+            _ => LogLevel.Debug
+        };
+
+        _sawmillNet.Log(logLevel, FormatLog(msg));
+        return Task.CompletedTask;
+    }
+
+    private Task OnCommandReceivedInternal(SocketMessage message)
+    {
+        var content = message.Content;
+        // If the message is too short to be a command, or doesn't start with the bot prefix, ignore it.
+        if (content.Length <= BotPrefix.Length || !content.StartsWith(BotPrefix))
+            return Task.CompletedTask;
+
+        // Split the message into the command and the arguments.
+        var split = content[BotPrefix.Length..].Split(' ', StringSplitOptions.RemoveEmptyEntries);
```

Comment:
Doing argument splitting in such a simplistic manner makes it impossible to pass arguments that contain spaces. It seems like a bad limitation to have.

Reply:
The arguments array is just meant to make simple commands easier. Anything more complicated can call upon the message object to get the raw message. I think this is fine to keep.

Reply:
I would recommend having the same system as console commands in that case, where everything *after* the command prefix is given. So commands don't have to do the work of stripping the prefix themselves.